mava/systems/sac/ff_masac.py

```diff
@@ -0,0 +1,638 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import time
+from typing import Any, Callable, Dict, Tuple
+
+import chex
+import flashbax as fbx
+import hydra
+import jax
+import jax.lax as lax
+import jax.numpy as jnp
+import optax
+from colorama import Fore, Style
+from flashbax.buffers.flat_buffer import TrajectoryBuffer
+from flax.core.scope import FrozenVariableDict
+from jax import Array
+from jumanji.env import Environment, State
+from omegaconf import DictConfig, OmegaConf
+from rich.pretty import pprint
+
+from mava.evaluator import make_eval_fns
+from mava.networks import FeedForwardActor as Actor
+from mava.networks import FeedForwardQNet as QNetwork
+from mava.systems.sac.types import (
+    BufferState,
+    LearnerState,
+    Metrics,
+    Networks,
+    Optimisers,
+    OptStates,
+    QVals,
+    QValsAndTarget,
+    SacParams,
+    Transition,
+)
+from mava.types import ObservationGlobalState
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.jax import unreplicate_batch_dim, unreplicate_n_dims
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.utils.total_timestep_checker import check_total_timesteps
+from mava.wrappers import episode_metrics
+
+
+def get_joint_action(actions: Array) -> Array:
```

Comment:
It will be better if we can have this in utils or somewhere else.
Also if you can add docs to mention the motivation behind having this as it is needed for shared action info in the Qs calculation